core/src/main/java/org/apache/iceberg/encryption/KeyMetadataDecoder.java

```diff
@@ -0,0 +1,145 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.message.BadHeaderException;
+import org.apache.avro.message.MessageDecoder;
+import org.apache.avro.message.MissingSchemaException;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.avro.ProjectionDatumReader;
+import org.apache.iceberg.data.avro.DataReader;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.MapMaker;
+
+public class KeyMetadataDecoder<D> extends MessageDecoder.BaseDecoder<D> {
+  private final org.apache.iceberg.Schema readSchema;
+  private final Map<Byte, RawDecoder<D>> decoders = new MapMaker().makeMap();
+  private final byte[] versionBuffer;
+
+  /**
+   * Creates a new decoder that constructs datum instances described by an {@link
+   * org.apache.iceberg.Schema Iceberg schema}.
+   *
+   * <p>The {@code readSchema} is as used the expected schema (read schema). Datum instances created
+   * by this class will are described by the expected schema.
+   *
+   * @param readSchema the schema used to construct datum instances
+   */
+  public KeyMetadataDecoder(byte currentSchemaVersion, org.apache.iceberg.Schema readSchema) {
+    this.readSchema = readSchema;
+    addSchema(currentSchemaVersion, this.readSchema);
+    this.versionBuffer = new byte[1];
+  }
+
+  /**
+   * Adds an {@link org.apache.iceberg.Schema Iceberg schema} that can be used to decode buffers.
+   *
+   * @param writeSchema a schema to use when decoding buffers
+   */
+  public void addSchema(byte schemaVersion, org.apache.iceberg.Schema writeSchema) {
+    decoders.put(
+        schemaVersion,
+        new KeyMetadataDecoder.RawDecoder<>(
+            readSchema, AvroSchemaUtil.convert(writeSchema, "key_metadata")));
+  }
+
+  @Override
+  public D decode(InputStream stream, D reuse) throws IOException {
+    try {
+      if (!readFully(stream, versionBuffer)) {
+        throw new BadHeaderException("Failed to read version - no bytes in stream");
+      }
+    } catch (IOException e) {
+      throw new IOException("Failed to read the version byte", e);
+    }
+
+    byte writeSchemaVersion = versionBuffer[0];
+
+    RawDecoder<D> decoder = decoders.get(writeSchemaVersion);
+
+    if (decoder == null) {
+      throw new MissingSchemaException("Cannot resolve schema for version: " + writeSchemaVersion);
+    }
+
+    return decoder.decode(stream, reuse);
+  }
+
+  private static class RawDecoder<D> extends BaseDecoder<D> {
```

Comment:
Why did this remove the threadlocal decoders?

If possible, I wanted this PR to refactor and share code with the `IcebergDecoder`. I don't see why `RawDecoder` was changed, or why it was copied rather than moved. Can you move it instead?